README
Cached Values
==============

A dead-simple way to calculate any value via Ruby or SQL and (optionally) have it saved in a database field.

Example
=======

USAGE:

You can calculate values with a single line in any ActiveRecord model.  To actually save those values you'll need to create
an attribute in your schema.  By default the cached_value instance will look for an attribute with the same name as itself.
You can override this by specifying a :cache => :some_attribute option

A very simple case in which cached_values works just like the .count method on a has_many association:

  class Leprechaun < ActiveRecord::Base
   caches_value :total_gold_coins, :sql => 'select count(*) from gold_coins where leprechaun_id = #{id}'
  end
  
  Company.find(4).total_employees # => 45

A more sophisticated example:

  class Leprechaun < ActiveRecord::Base
   has_many :lucky_charms
   has_many :deceived_children, :through => :lucky_charms
   caches_value :total_children_remaining_to_be_deceived, :sql => '... very complicated sql here ...'
  end
  
  Leprechaun.find(14).total_children_remaining_to_be_deceived # => 6,692,243,122

The values can be of any type.  The plugin will attempt to cast SQL results to the type corresponding with their database cache
but calculations in Ruby are left alone.

You can also calculate the value in Ruby using a string to be eval'ed or a Proc.  Both are evaluated
in the context of the record instance.

  class Leprechaun < ActiveRecord::Base
    caches_value :total_gold, :eval => "some_archaic_and_silly_calculation(self.gold_coins)"
    caches_value :total_lucky_charms, :eval => Proc.new {|record| record.calculate_total_lucky_charms }
  end

The cache is customizable, you can specify which attribute should be used as a cache:
  
  caches_value :runic_formula, :sql => '...'          # uses 'full_formula' column if it exists
  caches_value :standard_deviation_of_gold_over_time, # uses 'std' column if it exists
               :sql => '...', :cache => 'std'
  caches_value :id, :sql => '...', :cache => false    # does NOT save any cache.  This avoids overwriting an attribute

ActiveRecord callbacks can be used to call load (make sure the value is in memory), clear (flush the cache and
delete the instance), or reload (flush cache and reload instance) at certain times:
  
  caches_value :standard_deviation, :sql => '...', :reload => [:after_save, :before_validation]
  caches_value :full_formula, :sql => '...', :clear => :after_find

Bugs can be filed at http://code.lighthouseapp.com/projects/4502-hoopla-rails-plugins/

Copyright (c) 2007 Jack Danger Canty @ http://6brand.com, released under the MIT license


History.txt
=== 1.0.1 / 2008-03-12

* The find_target method is now public

=== 1.0.0 / 2008-03-06

* Converted from a Rails plugin


test/schema.rb

require File.expand_path(File.dirname(__FILE__) + "/leprechaun")

ActiveRecord::Schema.define(:version => 1) do
  
  create_table :leprechauns do |t|
    t.column :name,                                             :string
    t.column :favorite_color,                                   :string
    t.column :favorite_color_in_rot_13,                         :string
    t.column :favorite_color_in_rot_13_without_cache,           :string
    t.column :favorite_color_turned_uppercase,                  :string
    t.column :id_of_first_leprechaun_with_same_favorite_color,  :integer
    t.column :some_other_cache_field,                           :string
    t.column :reload_callback,                                  :float
    t.column :clear_callback,                                   :float
  end
  
  Leprechaun.create!(:name => 'Mc Nairn', :favorite_color => 'blue')
  Leprechaun.create!(:name => "O' Houhlihan", :favorite_color => 'ochre')
  Leprechaun.create!(:name => "Danger Cáinte", :favorite_color => 'blue')
  
end


test_helper.rb
# Include this file in your test by copying the following line to your test:
#   require File.expand_path(File.dirname(__FILE__) + "/test_helper")

$:.unshift(File.dirname(__FILE__) + '/../lib')

require 'rubygems'
require 'test/unit'
require 'active_record'
require 'active_record/fixtures'
require "cached_values"
require File.expand_path(File.dirname(__FILE__) + "/test/leprechaun")

config = YAML::load(IO.read(File.dirname(__FILE__) + '/test/database.yml'))
ActiveRecord::Base.logger = Logger.new(File.dirname(__FILE__) + "/test/debug.log")
ActiveRecord::Base.establish_connection(config[ENV['DB'] || 'sqlite3'])

load(File.dirname(__FILE__) + "/schema.rb") if File.exist?(File.dirname(__FILE__) + "/test/schema.rb")

class Test::Unit::TestCase #:nodoc:
  # def create_fixtures(*table_names)
  #   if block_given?
  #     Fixtures.create_fixtures(Test::Unit::TestCase.fixture_path, table_names) { yield }
  #   else
  #     Fixtures.create_fixtures(Test::Unit::TestCase.fixture_path, table_names)
  #   end
  # end

  # Turn off transactional fixtures if you're working with MyISAM tables in MySQL
  self.use_transactional_fixtures = true
  
  # Instantiated fixtures are slow, but give you @david where you otherwise would need people(:david)
  self.use_instantiated_fixtures  = false

  # Add more helper methods to be used by all tests here...
end

test/leprechaun.rb
class Leprechaun < ActiveRecord::Base
  caches_value :favorite_color_in_rot_13, :eval => Proc.new {|leprechaun| leprechaun.favorite_color.tr "A-Za-z", "N-ZA-Mn-za-m" }
  caches_value :favorite_color_turned_uppercase, :eval => "favorite_color.upcase"
  caches_value :number_of_gold_coins, :eval => :calculate_gold
  caches_value :id_of_first_leprechaun_with_same_favorite_color, :sql => 'select id from leprechauns where leprechauns.favorite_color = "#{favorite_color}" and leprechauns.id <> #{id} limit 1'
  caches_value :favorite_color_in_rot_13_without_cache, :eval => Proc.new {|leprechaun| leprechaun.favorite_color.tr "A-Za-z", "N-ZA-Mn-za-m" }, :cache => false
  caches_value :favorite_color_turned_uppercase_with_explicit_cache, :eval => "favorite_color.upcase", :cache => 'some_other_cache_field'
  caches_value :reload_callback, :eval => "rand(1000)", :reload => [:before_save, :after_validation]
  caches_value :clear_callback, :eval => "rand(1000)", :clear => :before_validation
  caches_value :float_cast, :sql => "select 82343.222"
  caches_value :integer_cast, :sql => "select 19"
  caches_value :string_cast, :sql => 'select "Top \'o the mornin\' to ya"'
  
  def calculate_gold
    '127 gold coins'
  end
end


lib/has_cached_value.rb
module HasCachedValueExtension # :nodoc:
  def self.included(base)
    base.extend(ClassMethods)
  end
  module ClassMethods
    # USAGE:
    #  
    # a very simple case in which has_cached_value works just like the .count method on a has_many association:
    #   
    #   class Company < ActiveRecord::Base
    #     has_cached_value :total_employees, :sql => 'select count(*) from employees where company_id = #{id}'
    #   end
    #  
    # a more sophisticated example:
    #  
    #   class User < ActiveRecord::Base
    #     has_many :trinkets
    #     has_many :sales, :through => :trinkets
    #     has_cached_value :remaining_trinket_sales_allotted, :sql => '... very complicated sql here ...'
    #   end
    #  
    #   user = User.find(:first)
    #   user.remaining_trinket_sales_allotted # => 70
    #   Trinket.delete_all # <= any operation that would affect our value
    #   user.remaining_trinket_sales_allotted # => 70
    #   user.remaining_trinket_sales_allotted.reload # => 113
    #  
    # You can also calculate the value in Ruby.  This can be done by a string to be eval'ed or a Proc.  Both are evaluated
    # in the context of the record instance.
    #  
    #   class User < ActiveRecord::Base
    #     has_cached_value :expensive_calculation, :eval => "some_big_expensize_calculation(self.id)"
    #     has_cached_value :other_expensive_process, :eval => Proc.new {|record| record.other_expensize_process }
    #   end
    #
    
    def has_cached_value(association_id, options = {})
      reflection = create_has_cached_value_reflection(association_id, options)

      configure_dependency_for_has_cached_value(reflection)

      reflection.options[:counter_cache] = reflection.options.delete(:cache) if reflection.options[:cache]
      reflection.options[:counter_cache] ||= reflection.name unless false == reflection.options[:counter_cache]

      association_accessor_method(reflection, ActiveRecord::Associations::HasCachedValueAssociation)
    end

    private
    
      def configure_dependency_for_has_cached_value(reflection)
        if reflection.options[:counter_cache] && reflection.options[:cache]
          raise ArgumentError, ":cache is an alias for :counter_cache, don't use both options for has_cached_value in #{self.name}"
        end
      
        if !reflection.options[:sql] && !reflection.options[:eval]
          raise ArgumentError, "You must specify either the :eval or :sql options for has_cached_value in #{self.name}"
        end
      
        if reflection.options[:sql] && reflection.options[:eval]
          raise ArgumentError, ":eval and :sql are mutually exclusive options.  You may specify one or the other for has_cached_value in #{self.name}"
        end
      end
    
      def create_has_cached_value_reflection(association_id, options)
        options.assert_valid_keys(:sql, :eval, :cache, :counter_cache)
        
        reflection = ActiveRecord::Reflection::AssociationReflection.new(:has_cached_value, association_id, options, self)
        write_inheritable_hash :reflections, association_id => reflection
        reflection
      end

      def association_accessor_method(reflection, association_proxy_class)
        define_method(reflection.name) do |*params|
          force_reload = params.first unless params.empty?
          association = instance_variable_get("@#{reflection.name}")
          
          if association.nil? || force_reload
            association = association_proxy_class.new(self, reflection)
            instance_variable_set("@#{reflection.name}", association)
            force_reload ? association.reload : association.load
          end
          association.target.nil? ? nil : association
        end
      end
  end
end

module ActiveRecord
  module Associations
    class HasCachedValueAssociation
      attr_reader :reflection
      instance_methods.each { |m| undef_method m unless m =~ /(^__|^nil\?$|^send$)/ }

      def initialize(owner, reflection)
        @owner, @reflection = owner, reflection
        reset
      end
      
      def reset
        @target = nil
        @loaded = false
      end

      def load
        reset
        load_target
      end

      def reload
        reset
        clear_cache
        load_target
      end
      
      def clear
        clear_cache
        @owner.instance_variable_set("@#{@reflection.name}", nil)
      end

      def loaded?
        @loaded
      end
      
      def loaded
        @loaded = true
      end
      
      def target
        @target
      end

      protected
      
        def find_target_by_eval
          if @reflection.options[:eval].is_a?(String)
            eval(@reflection.options[:eval], @owner.send(:binding))
          elsif @reflection.options[:eval].is_a?(Proc)
            @reflection.options[:eval].call(@owner)
          else
            raise ArgumentError.new("The :eval option on a has_cached_value must be either a String or a Proc")
          end
        end
        
        def find_target_by_sql
          @owner.class.count_by_sql(sanitize_sql(interpolate_sql(@reflection.options[:sql])))
        end
        
        def find_target_from_cache
          @owner.send(:read_attribute, @reflection.counter_cache_column) if has_cached_counter?
        end

        def find_target
          target = find_target_from_cache
          unless target
            target ||= @reflection.options[:sql] ? find_target_by_sql : find_target_by_eval
            update_cache(target)
          end
          target
        end
        
        def update_cache(value)
          return unless has_cached_counter?
          unless @owner.new_record?
            @owner.class.update_all(["#{@reflection.counter_cache_column} = ?", value], ["id = ?", @owner.id])
          end
          @owner.send(:write_attribute, @reflection.counter_cache_column, value)
        end
        
        def clear_cache
          update_cache(nil)
        end
        
        def load_target
          return nil unless defined?(@loaded)
          @target = find_target unless loaded?
          @loaded = true
          @target
        end

        def has_cached_counter?
          @reflection.options[:counter_cache] && @owner.attribute_names.include?(@reflection.options[:counter_cache].to_s)
        end
        
        def interpolate_sql(sql, record = nil)
          @owner.send(:interpolate_sql, sql, record)
        end

        def sanitize_sql(sql)
          @owner.class.send(:sanitize_sql, sql)
        end
        
        def method_missing(method, *args, &block)
          if load_target        
            @target.send(method, *args, &block)
          end
        end
    end
  end
end


lib/cached_values/cached_value.rb
require 'object_proxy'
  
module ActiveRecord
  class CachedValue < ObjectProxy

    def initialize(owner, reflection)
      @owner, @reflection = owner, reflection
      reset
    end
    
    def reset
      @target = nil
    end

    def load
      reset
      @target = find_target(true)
      update_cache(@target)
    end

    def reload
      @owner.instance_variable_set("@#{@reflection.name}", nil)
      load
      @owner.send @reflection.name
    end
    
    alias update reload
    
    def clear
      clear_cache
      @owner.instance_variable_set("@#{@reflection.name}", nil)
    end

    def target
      @target
    end

    def find_target(skip_cache = false)
      target = find_target_from_cache unless skip_cache
      unless target
        target ||= @reflection.options[:sql] ? find_target_by_sql : find_target_by_eval
      end
      target
    end
    
    protected
      
      def find_target_from_cache
        @owner.send(:read_attribute, cache_column) if has_cache?
      end
      
      def find_target_by_sql
        sql = sanitize_sql(interpolate_sql(@reflection.options[:sql]))
        result = @owner.class.connection.select_value(sql)
        result = typecast_result(result)
        result
      end
    
      def find_target_by_eval
        if @reflection.options[:eval].is_a?(String)
          eval(@reflection.options[:eval], @owner.send(:binding))
        elsif @reflection.options[:eval].is_a?(Proc)
          @reflection.options[:eval].call(@owner)
        elsif @reflection.options[:eval].is_a?(Symbol)
          @owner.send @reflection.options[:eval]
        else
          raise ArgumentError.new("The :eval option on a cached_values must be a String or a Proc or a Symbol")
        end
      end
      
      def cache_column
        @reflection.options[:cache]
      end

      def has_cache?
        @reflection.options[:cache] && @owner.attribute_names.include?(@reflection.options[:cache].to_s)
      end
      
      def clear_cache
        update_cache(nil)
      end
      
      def update_cache(value)
        return unless has_cache?
        unless @owner.new_record?
          @owner.class.update_all(["#{cache_column} = ?", value], ["id = ?", @owner.id])
        end
        @owner.send(:write_attribute, cache_column, value) unless @owner.frozen?
      end
      
      def typecast_result(result)
        if has_cache?
          typecast_sql_result_by_cache_column_type(result)
        else
          if result =~ /^\d+\.\d+$/
            result.to_f
          elsif result =~ /^\d+$/
            result.to_i
          else
            result
          end
        end
      end
      
      def typecast_sql_result_by_cache_column_type(result)
        type = @owner.column_for_attribute(cache_column).type
        case type
        when :integer
          result.to_i
        when :float
          result.to_f
        when :boolean
          [0,'0', 'NULL', 'nil', 'false', '', 'FALSE', 'False'].include?(result)
        else
          result
        end
      end
      
      def interpolate_sql(sql, record = nil)
        @owner.send(:interpolate_sql, sql, record)
      end

      def sanitize_sql(sql)
        @owner.class.send(:sanitize_sql, sql)
      end
      
  end
end


lib/cached_values/caches_value.rb
require 'active_record'
require File.expand_path(File.dirname(__FILE__) + '/cached_value')

module CachesValues # :nodoc:
  
  def self.included(base)
    unless base.extended_by.include?(CachesValues::ClassMethods)
      base.extend(ClassMethods)
    end
  end

  module ClassMethods
    # USAGE:
    #  
    # a very simple case in which cached_values works just like the .count method on a has_many association:
    #   
    #   class Company < ActiveRecord::Base
    #     caches_value :total_employees, :sql => 'select count(*) from employees where company_id = #{id}'
    #   end
    #  
    # a more sophisticated example:
    #  
    #   class User < ActiveRecord::Base
    #     has_many :trinkets
    #     has_many :sales, :through => :trinkets
    #     caches_value :remaining_trinket_sales_allotted, :sql => '... very complicated sql here ...'
    #   end
    #  
    #   user = User.find(:first)
    #   user.remaining_trinket_sales_allotted # => 70
    #   Trinket.delete_all # <= any operation that would affect our value
    #   user.remaining_trinket_sales_allotted # => 70
    #   user.remaining_trinket_sales_allotted.reload # => 113
    #  
    # You can also calculate the value in Ruby.  This can be done by a string to be eval'ed or a Proc.  Both are evaluated
    # in the context of the record instance.
    #  
    #   class User < ActiveRecord::Base
    #     caches_value :expensive_calculation, :eval => "some_big_expensize_calculation(self.id)"
    #     caches_value :other_expensive_process, :eval => Proc.new {|record| record.other_expensize_process }
    #   end
    #
    
    def caches_value(name, options = {})
      reflection = create_cached_value_reflection(name, options)

      configure_dependency_for_cached_value(reflection)

      reflection.options[:cache] ||= reflection.name unless false == options[:cache]

      cached_value_accessor_method(reflection, ActiveRecord::CachedValue)
      cached_value_callback_methods(reflection)
    end

    private
    
      def configure_dependency_for_cached_value(reflection)
      
        if !reflection.options[:sql] && !reflection.options[:eval]
          raise ArgumentError, "You must specify either the :eval or :sql options for caches_value in #{self.name}"
        end
      
        if reflection.options[:sql] && reflection.options[:eval]
          raise ArgumentError, ":eval and :sql are mutually exclusive options.  You may specify one or the other for caches_value in #{self.name}"
        end
      end
    
      def create_cached_value_reflection(name, options)
        options.assert_valid_keys(:sql, :eval, :cache, :clear, :load, :reload)
        
        reflection = ActiveRecord::Reflection::MacroReflection.new(:cached_value, name, options, self)
        write_inheritable_hash :reflections, name => reflection
        reflection
      end

      def cached_value_accessor_method(reflection, association_proxy_class)
        define_method(reflection.name) do |*params|
          force_reload = params.first unless params.empty?
          association = instance_variable_get("@#{reflection.name}")
          
          if association.nil? || force_reload
            association = association_proxy_class.new(self, reflection)
            instance_variable_set("@#{reflection.name}", association)
            force_reload ? association.reload : association.load
          end
          association.target.nil? ? nil : association
        end
      end
      
      def cached_value_callback_methods(reflection)
        %w{clear reload}.each do |operation|
          if events = reflection.options[operation.to_sym]
            events = [events] unless events.is_a?(Array)
            events.map! { |event| event.to_s }
            ActiveRecord::Callbacks::CALLBACKS.each do |callback|
              if events.include?(callback)
                send(callback, Proc.new {|record| record.send(reflection.name).send(operation)})
              end
            end
          end
        end
      end
  end
end
